lib/factories/default-eventgate.js

```diff
@@ -168,34 +183,135 @@ function makeValidate(options, logger) {
+    );
+
+    // If rdkafka is configured with statistics reporting turned on,
+    // AND if we are given a metrics object (node-statsd like or from service-runner app directly)
+    // then enable Kafka metrics reporting using node-rdkafka-statsd.
+    if (options.kafka.conf['statistics.interval.ms'] && metrics) {
+        const rdkafkaStatsd  = require('node-rdkafka-statsd');
+        logger.info('Enabling Kafka metrics reporting for Kafka producer');
+
+        kafkaProducer.on(
+            'event.stats',
+            rdkafkaStatsd(metrics.makeChild('rdkafka.producer'))
+        );
+    }
+
     // Return a new function that takes a single event argument for produce.
     return (event, context = {}) => {
         const topic = extractTopic(event);
         const serializedEvent = Buffer.from(JSON.stringify(event));
-        return producer.produce(topic, undefined, serializedEvent, undefined);
+        return kafkaProducer.produce(topic, undefined, serializedEvent, undefined);
+    };
+}
+
+
+/**
+ * Creates a function that will write events to an output file and/or
+ * produce events to Kafka.
+ *
+ * @param {Object} options
+ * @param {string} options.stream_field
+ *      Used to construct the kafka topic
+* @param {string} options.schema_uri_field
+ *      Used to construct the kafka topic if stream_field is not set.
+ * @param {string} options.key_field
+ *      If given, the event's 'key' will be extracted from the event at this dotted object path.
+ * @param {string} options.partition_field
+ *      If given, the event's 'partition' will be extracted from the event at this
+ *      dotted object path.
+ * @param {Object} options.output_path
+ *      If given, events will be written to this file.  If 'stdout' events will be
+ *      written to stdout.
+ * @param {Object} options.kafka.conf
+ *      If given, events will be produced to Kafka.
+ *      node-rdkafka KafkaProducer configuration
+ * @param {Object} options.kafka.topic_conf
+ *      node-rdkafka KafkaProducer topic configuration
+ * @param {Object} logger
+ * @param {Object} metrics
+ *      service-runner metrics object.
+ * @return {Promise<{EventGate~produce}>} (event, context) => Promise<event>>}
+ */
+async function makeProduce(options, logger, metrics) {
+
+    let fileProduce;
+    if (options.output_path) {
+        fileProduce = makeFileProduce(options, logger, metrics);
+        logger.info('Writing valid events to ' + options.output_path);
+    }
+
+    let kafkaProduce;
+    if (options.kafka && options.kafka.conf) {
+        kafkaProduce = await makeKafkaProduce(options, logger, metrics);
+        logger.info('Producing valid events to Kafka at ' + options.kafka.conf['metadata.broker.list']);
+    }
+
+    // Return a new function that takes a single event argument for produce.
+    return async (event, context = {}) => {
+        if (fileProduce) {
+            await fileProduce(event, context);
+            await kafkaProduce(event, context);
+        }
```

Comment:
This is clearly quite wrong.. what if we are only producing to a file? there should be 2 ifs

Reply:
OO good catch thank youl.